Environment Set-Up (#56)
hytest-org/hytest: environment_set_up/ManualConfig-HPC.md

# Getting Started with On-Prem HPC

This tutorial covers how to set up and use the Pangeo environment/stack
on  USGS High Performance Computing (HPC) systems:

1) Creating a custom hytest conda environment
2) Configuring [Jupyter](https://jupyter.org/) server
3) Launching a Jupyter server
4) Connecting to your Jupyter server and the[Dask](https://dask.pydata.org/) 
   dashboard from your personal computer

This document assumes that you already have an access to
Denali and Tallgrass and are comfortable using the command line.

> ## Installing a Conda Software Environment

Reviewer: Can we name this the same thing between the two files?

Author: I'm a little lost... what's "this" ? to name the same between which two files?

Reviewer: Oh the section title...the text below says "This is the same manual process as describe in the 
[Quick-Start for HPC](./QuickStart-HPC.md)" but I don't see a section with that naming in the file it links to. Where do I find the instructions to install a conda software environment?

Author: I made a link directly to the header/anchor for the conda setup step.

> ## Installing a Conda Software Environment

Reviewer: I think this is what we want?
```suggestion
## Set up a `hytest` Conda environment
```

Author: done

This is the same manual process as describe in the 

> [Quick-Start for HPC](../QuickStart-HPC.md)

Reviewer: ```suggestion
[Quick-Start for HPC](./QuickStart-HPC.md)
```

## Configure Jupyter

Jupyter notebook servers include a password for security. First we generate the Jupyter config file then set a password:

```text
jupyter server --generate-config
jupyter server password
```

This creates the file `~/.jupyter/jupyter_server_config.py`.

Finally, we configure dask\'s dashboard to forward through
Jupyter.  Add the following line to your `~/.bashrc` file:

```sh
export DASK_DISTRIBUTED__DASHBOARD__LINK="/proxy/8787/status"
```

## Start a Jupyter Notebook Server

Now that we have Jupyter configured, we can start a notebook server on our interactive compute node. Use the [script provided](./start_jupyter.sh) to do this.

```text
> salloc -A account_name -t 02:00:00 -N 1 srun --pty bash

> > cd $HOME/HyTEST-Tutorials/
> > ./HPC/start_jupyter.sh

Reviewer: The cd line needs to change too if we aren't copying in that zip folder...just making a note to come back once I figure out what it should be instead.
```suggestion
> cd $HOME/HyTEST-Tutorials/
> ./environment_set_up/start_jupyter.sh
```

```

Follow the steps printed out by the script to get connected. Note that after you login using the `ssh` command in Step 2, the cursor will hang.  This is expected and you can minimize the window.

## Running Notebooks

> ### Special Notes for dask jobqueue

Reviewer: I haven't used dask-jobqueue before so this is an uninformed question. From reading [Rich's instructions](https://github.com/USGS-python/hytest-development/blob/main/hpc.md#launch-dask-with-dask-jobqueue) it sounds like I have to run this Python code to get started.

I see that you are linking to a whole notebook with instructional material.

How does this code fit in to the workflow?
- Is it included in our notebooks and so telling the user to run this code is not necessary. And your notebook is just an explainer for what dask-jobqueue is?
- Or does this piece of code need to get run just once during setup?

Author: The original plan is to point to the jobqueue notebook as an explainer/demo.  That code actually gets executed as part of the `configure_cluster()` function, and would run automatically if a user uses that function.

Author: I wonder if we even need this link, actually.  Could we just strike that last sentence and reference to the xx_dask_jobqueue.ipynb notebook. ??

Reviewer: Thanks for clarifying - maybe we can move this into a `tutorial` subfolder of `environment_set_up` then and update any links to reflect the new path.

Reviewer: Yeah I think we should rephrase this and maybe just have text that says something like "you can check out the [xx_dask-jobqueue.ipynb]() notebook for more info"

Reviewer 2: We use dask-jobqueue when we are on tallgrass, because there are only a few CPUs per node and we are often sharing them with others.

Author: I changed the language a little -- and I converted the jobqueue notebook to a flat markdown file.  Seems like that document just contained example code, rather than something that we want a reader to execute within that document.

Most HPC systems use a job-scheduling system to manage job submissions
and executions among many users.
The [dask-jobqueue](http://dask-jobqueue.readthedocs.io) package is designed to help
dask interface with these job queuing systems. Usage is quite simple and can be done
from within your Jupyter notebook. See

> [this doc](../Syllabus/L2/xx_dask-jobqueue.ipynb) for dask-specific details and how

Reviewer: ```suggestion
[this notebook](./xx_dask-jobqueue.ipynb) for dask-specific details and how
```

Author: thanks

> [this doc](../Syllabus/L2/xx_dask-jobqueue.ipynb) for dask-specific details and how

Reviewer: update link here

Author: I'll add the dask-jobqueue.ipynb as a commit to this branch.  Stand by.

Author: I have a jobqueue notebook committed now.  It is almost a direct copy of the original, and seems very specific to tallgrass.  I am not able to test anything (yet), so I am losing confidence in it.

# Getting Started with On-Prem HPC

This tutorial covers how to set up and use the Pangeo environment/stack

> on  USGS High Performance Computing (HPC) systems:

Reviewer: ```suggestion
on USGS High Performance Computing (HPC) systems:
```

1) Creating a custom `hytest` conda environment
2) Configuring [Jupyter](https://jupyter.org/) server
3) Launching a Jupyter server
4) Connecting to your Jupyter server and the [Dask](https://dask.pydata.org/)
   dashboard from your personal computer

This document assumes that you already have an access to

> `denali` or `tallgrass` and are comfortable using the command line.

Reviewer: ```suggestion
either the `Denali` or `Tallgrass` [supercomputers at USGS](https://hpcportal.cr.usgs.gov/hpc-user-docs/index.html) and are comfortable using the command line.
```

## Set up a `hytest` Conda Environment

This is the same manual process as describe in the
[Quick-Start for HPC](./QuickStart-HPC.md#1b-set-up-a-hytest-conda-environment)
to set up a `hytest` conda environment.

## Configure Jupyter

Jupyter notebook servers include a password for security. This will control who
can use your personal jupyter server on the HPC. First we
generate the Jupyter config file then set the password:

```text
jupyter server --generate-config
jupyter server password
```

This creates the file `~/.jupyter/jupyter_server_config.py`.

Finally, we configure dask's dashboard to forward through
Jupyter.  Add the following line to your `~/.bashrc` file:

```sh
export DASK_DISTRIBUTED__DASHBOARD__LINK="/proxy/8787/status"
```

## Start a Jupyter Notebook Server

Now that we have Jupyter configured, we can start a notebook server on our interactive compute node. Use the [script provided](start_jupyter.sh) to do this.

```text
> salloc -A account_name -t 02:00:00 -N 1 srun --pty bash
> cd $HOME/hytest/
> ./environment_set_up/start_jupyter.sh
```

> Follow the steps printed out by the script to get connected. Note that after you login using the `ssh` command in Step 2, the cursor will hang.  This is expected and you can minimize the window.

Reviewer: I don't see an `ssh` command in step 2 (start a jupyter notebook server)

Author: This script and example are taken from 
https://github.com/USGS-python/hytest-development/blob/main/hpc.md

And because I'm not yet on the HPC, I can't test.  So I might need to call in some help on that. 

I think that it would be useful to rename 'Step' in the script output so that it isn't confusing with the steps in the ManualConfig doc.

> ## Running Notebooks

Reviewer: does this section still need to have text added?

Author: I added a couple of lines... users can now find notebooks (or create them) in their HPC account file system.

### Special Notes for dask jobqueue

> The `tallgrass` node uses a job-scheduling system to manage job submissions
> and executions among many users. The [dask-jobqueue](http://dask-jobqueue.readthedocs.io)
> package is designed to help dask interface with these job queuing systems.
> 
> See [here](tallgrass_dask-jobqueue.md) for dask-specific details and how
> to interact with `dask-jobqueue` from within your notebook if it will be running on `tallgrass`.

Reviewer: ```suggestion
The `tallgrass` node uses a job-scheduling system to manage job submissions
and executions among many users. The [dask-jobqueue](http://dask-jobqueue.readthedocs.io)
package is designed to help dask interface with these job queuing systems. Our notebooks are set up to handle this task for you, but if you would like to learn more, see [here](tallgrass_dask-jobqueue.md) for dask-specific details and how
to interact with `dask-jobqueue` from within your notebook if it will be running on `tallgrass`.
```

# Getting Started with On-Prem HPC

This tutorial covers how to set up and use the Pangeo environment/stack
on USGS High Performance Computing (HPC) systems:

1) Creating a custom `hytest` conda environment
2) Configuring [Jupyter](https://jupyter.org/) server
3) Launching a Jupyter server
4) Connecting to your Jupyter server and the [Dask](https://dask.pydata.org/)
   dashboard from your personal computer

This document assumes that you already have an access to
either the `Denali` or `Tallgrass` [supercomputers at USGS](https://hpcportal.cr.usgs.gov/hpc-user-docs/index.html) and are comfortable using the command line.

> ## Set up a `hytest` Conda Environment

Reviewer: ```suggestion
## 1) Set up a `hytest` Conda Environment
```

This is the same manual process as describe in the
[Quick-Start for HPC](./QuickStart-HPC.md#1b-set-up-a-hytest-conda-environment)
to set up a `hytest` conda environment.

> ## Configure Jupyter

Reviewer: ```suggestion
## 2) Configure Jupyter
```

Jupyter notebook servers include a password for security. This will control who
can use your personal jupyter server on the HPC. First we
generate the Jupyter config file then set the password:

```text
jupyter server --generate-config
jupyter server password
```

This creates the file `~/.jupyter/jupyter_server_config.py`.

Finally, we configure dask's dashboard to forward through
Jupyter.  Add the following line to your `~/.bashrc` file:

```sh
export DASK_DISTRIBUTED__DASHBOARD__LINK="/proxy/8787/status"
```

> ## Start a Jupyter Notebook Server

Reviewer: ```suggestion
## 3) Start a Jupyter Notebook Server
```

Now that we have Jupyter configured, we can start a notebook server on our interactive compute node. Use the [script provided](start_jupyter.sh) to do this.

```text
> salloc -A account_name -t 02:00:00 -N 1 srun --pty bash
> cd $HOME/hytest/
> ./environment_set_up/start_jupyter.sh
  * Wait until this script says the Jupyter server has started.

  * Copy this ssh command into a terminal on your local computer:

        ssh -N -L 8889:denali:8471  username@clustername.cr.usgs.gov

  * Browse to http://localhost:8889 using your web browser on your local computer
```

> 

Reviewer: ```suggestion

`account_name` is one of your accounts codes on the supercomputer. You will see a list of your account codes when you first log in to the supercomputer. They are listed in a table after some text saying `Your account codes are:`.

```

Follow the steps printed out by the script to get connected.
Note that after you execute the `ssh` command, the cursor will hang.  This is expected and you can minimize the window.

## Running Notebooks

> Once the jupyter server is activated and accepting connections on the HPC node, you may use the web
> interface to navigate to and run notebooks within your account, or create new ones.

Reviewer: Stealing the text from below instead because it is a nice description. I'm not sure I linked the section right though...
```suggestion
You are now ready to connect a web browser to your jupyter server.  Launch
a web browser on your dekstop and connect to the URL provided in the
startup messaging when you launched `jupyter lab`.

> NOTE:  Don't use the `localhost` option.  Use the URL which includes
your HPC host's full network name. That should look something like
<kbd>https://denali:8402/</kbd>.

The Jupyter Server will ask for a password.  This is the one supplied
in [Step 2](#2-start-a-jupyter-notebook-server) above.

You can now run existing notebooks found in `hytest`, or create
your own.
```
